src/fileformat/file_format/pe/authenticode/x509_certificate.cpp

```diff
@@ -0,0 +1,273 @@
+		else if (key == "GN")
+			attributes.givenName = value;
+		else if (key == "initials")
+			attributes.initials = value;
+		else if (key == "pseudonym")
+			attributes.pseudonym = value;
+		else if (key == "generationQualifier")
+			attributes.generationQualifier = value;
+		else if (key == "emailAddress")
+			attributes.emailAddress = value;
+	}
+
+	return attributes;
+}
+
+Certificate::Attributes X509Certificate::getSubject() const
+{
+	return parseAttributes(X509_get_subject_name(cert));
+}
+Certificate::Attributes X509Certificate::getIssuer() const
+{
+	return parseAttributes(X509_get_issuer_name(cert));
+}
+
+std::string X509Certificate::getPublicKey() const
+{
+	std::uint8_t* data = nullptr;
+	EVP_PKEY* pkey = X509_get0_pubkey(cert);
+	BIO* memBio = BIO_new(BIO_s_mem());
+	PEM_write_bio_PUBKEY(memBio, pkey);
+
+	std::string result(parsePublicKey(memBio));
+	BIO_free_all(memBio);
+
+	return result;
+}
+
+std::string X509Certificate::getPublicKeyAlgorithm() const
+{
+	const EVP_PKEY* pkey = X509_get0_pubkey(cert);
+	if (!pkey) {
+		return "unknown";
+	}
+
+	return OBJ_nid2sn(EVP_PKEY_base_id(pkey));
+}
+
+std::string X509Certificate::getSha1() const
+{
+	const int sha1_length = 20;
+	std::uint8_t sha1_bytes[sha1_length];
+
+	std::uint8_t* data = nullptr;
+	int len = i2d_X509(const_cast<X509*>(cert), &data);
+
+	const EVP_MD* md = EVP_sha1();
+	calculateDigest(md, data, len, sha1_bytes);
+
+	free(data);
+	return bytesToHexString(sha1_bytes, sha1_length);
+}
+std::string X509Certificate::getSha256() const
+{
+	const int sha256_length = 32;
+	std::uint8_t sha256_bytes[sha256_length];
+
+	std::uint8_t* data = nullptr;
+	int len = i2d_X509(const_cast<X509*>(cert), &data);
+
+	const EVP_MD* md = EVP_sha256();
+	calculateDigest(md, data, len, sha256_bytes);
+
+	free(data);
+	return bytesToHexString(sha256_bytes, sha256_length);
+}
+
+int X509Certificate::getVersion() const
+{
+	return X509_get_version(cert);
+}
+
+std::string X509Certificate::getRawSubject() const
+{
+	return X509NameToString(X509_get_subject_name(cert));
+}
+
+std::string X509Certificate::getRawIssuer() const
+{
+	return X509NameToString(X509_get_issuer_name(cert));
+}
+
+Certificate X509Certificate::createCertificate() const
+{
+	Certificate out_cert;
+	out_cert.issuerRaw = getRawIssuer();
+	out_cert.subjectRaw = getRawSubject();
+	out_cert.issuer = getIssuer();
+	out_cert.subject = getSubject();
+	out_cert.publicKey = getPublicKey();
+	out_cert.publicKeyAlgo = getPublicKeyAlgorithm();
+	out_cert.signatureAlgo = getSignatureAlgorithm();
+	out_cert.serialNumber = getSerialNumber();
+	out_cert.sha1Digest = getSha1();
+	out_cert.sha256Digest = getSha256();
+	out_cert.validSince = getValidSince();
+	out_cert.validUntil = getValidUntil();
+	return out_cert;
+}
+
+// /* It's assumed that the certificates are processed in correct
+//    order by callbacks set in the CertProcessor constructor */
+
+// static CertificateProcessor* getProcessor(X509_STORE_CTX* ctx)
```

Comment:
we generally don't want to have this kind of dead code in the source base. Is this helpful in any way? Could it be removed?